trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedLoopInvariant.java

```diff
@@ -0,0 +1,72 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.cacsl2boogietranslator.witness;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.eclipse.cdt.core.dom.ast.IASTNode;
+
+import de.uni_freiburg.informatik.ultimate.boogie.ast.AssertStatement;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.LoopInvariantSpecification;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.Statement;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.WhileStatement;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.base.IDispatcher;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.ExpressionResult;
+import de.uni_freiburg.informatik.ultimate.cdt.translation.implementation.result.ExpressionResultBuilder;
+import de.uni_freiburg.informatik.ultimate.core.lib.models.annotation.Check;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ILocation;
+import de.uni_freiburg.informatik.ultimate.util.datastructures.DataStructureUtils;
+
+public class ExtractedLoopInvariant extends ExtractedWitnessInvariant {
+	public ExtractedLoopInvariant(final String invariant, final Collection<String> nodeLabel, final IASTNode match) {
+		super(invariant, nodeLabel, match);
+	}
+
+	@Override
+	public ExpressionResult transform(final ILocation loc, final IDispatcher dispatcher,
+			final ExpressionResult expressionResult) {
+		final ExpressionResult invariantExprResult = instrument(loc, dispatcher);
+		final List<Statement> statements = new ArrayList<>();
+		boolean hasLoop = false;
+		for (final Statement st : expressionResult.getStatements()) {
+			if (st instanceof WhileStatement) {
+				assert !hasLoop;
+				hasLoop = true;
+				final WhileStatement whileOld = (WhileStatement) st;
+				statements.add(new WhileStatement(loc, whileOld.getCondition(), DataStructureUtils
+						.concat(whileOld.getInvariants(), extractLoopInvariants(invariantExprResult, loc)),
+						whileOld.getBody()));
+			} else {
+				statements.add(st);
+			}
+		}
+		if (hasLoop) {
+			return new ExpressionResultBuilder(expressionResult).resetStatements(statements).build();
+		}
+		// We might identify sth. that is not a loop (e.g. goto) as a loop.
+		// We cannot annotate it with a a LoopInvariantSpecification, so we just insert an assert in that case.
+		return new ExpressionResultBuilder(invariantExprResult).addAllIncludingLrValue(expressionResult).build();
+	}
+
+	private static LoopInvariantSpecification[] extractLoopInvariants(final ExpressionResult result,
+			final ILocation loc) {
+		final List<LoopInvariantSpecification> specs = new ArrayList<>();
```

Comment:
I don't fully understand what happens here:

- Could we not directly return the needed data structures, rather than retroactively extracting them?
- How hard would it be to support other statements here that are perhaps required for the subsequent `assert`? Would it make sense?

Reply:
First we create an assert of the loop invariant, then we extract the assertion and make it a loop invariant in Boogie. Alternatively we could also directly instrument an ACSL loop invariant, but I have not looked to that in detail and just kept thr previous functionality.